src/field/field.rs

```diff
@@ -486,3 +486,212 @@ macro_rules! test_square_root {
         }
     };
 }
+
+pub mod field_tests {
+    use std::io::Result;
+    use crate::{Field, biguint_to_field, field_to_biguint};
+    use num::{BigUint, Zero, One};
+
+    /// Generates a series of non-negative integers less than
+    /// `modulus` which cover a range of values and which will
+    /// generate lots of carries, especially at `word_bits` word
+    /// boundaries.
+    fn test_inputs(modulus: &num::BigUint, word_bits: usize) -> Vec<num::BigUint>
+    {
+        assert!(word_bits == 32 || word_bits == 64);
+        let modwords = (modulus.bits() + word_bits - 1) / word_bits;
+        // Start with basic set close to zero: 0 .. 10
+        const BIGGEST_SMALL: u32 = 10;
+        let smalls: Vec<_> = (0..BIGGEST_SMALL).map(|x| BigUint::from(x)).collect();
+        // ... and close to MAX: MAX - x
+        let word_max = (BigUint::one() << word_bits) - 1u32;
+        let bigs = smalls.iter().map(|x| &word_max - x).collect();
+        let one_words = [smalls, bigs].concat();
+        // For each of the one word inputs above, create a new one at word i.
+        // TODO: Create all possible `modwords` combinations of those
+        let multiple_words = (1..modwords).flat_map(|i| {
+            one_words.iter().map(|x| x << (word_bits * i)).collect::<Vec<BigUint>>()
+        }).collect();
+        let basic_inputs = [one_words, multiple_words].concat();
+
+        // Biggest value that will fit in `modwords` words
+        let maxval = (BigUint::one() << (modwords * word_bits)) - 1u32;
+        // Inputs 'difference from' maximum value
+        let diff_max = basic_inputs.iter()
+            .map(|x| &maxval - x)
+            .filter(|x| x < &modulus)
+            .collect();
+        // Inputs 'difference from' modulus value
+        let diff_mod = basic_inputs.iter()
+            .filter(|x| *x < &modulus && !x.is_zero())
+            .map(|x| modulus - x)
+            .collect();
+        let basics = basic_inputs.into_iter().filter(|x| x < &modulus).collect::<Vec<BigUint>>();
+        [basics, diff_max, diff_mod].concat()
+
+        // // There should be a nicer way to express the code above; something
+        // // like this (and removing collect() calls from diff_max and diff_mod):
+        // basic_inputs.into_iter()
+        //     .chain(diff_max)
+        //     .chain(diff_mod)
+        //     .filter(|x| x < &modulus)
+        //     .collect()
+    }
+
+
+    /// Apply the unary functions `op` and `expected_op`
+    /// coordinate-wise to the inputs from `test_inputs(modulus,
+    /// word_bits)` and panic if the two resulting vectors differ.
+    pub fn run_unaryop_test_cases<T, UnaryOp, ExpectedOp>(
+        modulus: &BigUint, word_bits: usize,
+        op: UnaryOp, expected_op: ExpectedOp)
+        -> Result<()>
+    where
+        T: Field,
```

Comment:
nit: we usually use `F` for Field type params (ditto for `run_binaryop_test_cases`)